### ISSUE TITLE
Privacy across OS accounts

### ISSUE THREAD

**Issue by levangongPayPal** [w3c/webauthn#96]
This is more of question than anything else.
Section 1 - Use Cases says:
“Additionally, privacy across WebAuthn Relying Parties must be maintained; scripts must not be able to detect any properties, or even the existence, of scoped credentials belonging to other WebAuthn Relying Parties."

From a privacy perspective, is there anything we can say or do to enable privacy-preserving measures between users sharing a device but using different OS accounts?
Could we define an (API) extension or more likely recommend an implementation approach that would allow to link a credential to an OS account?
This would help avoiding the leak of information later on (e.g. when Bob registers or authenticates to a particular web site, he notices that Alice has also an account at that provider).
In other words, should we suggest that, whenever possible, credentials should be scoped to authenticator-OS account-RP or something like that?

**Comment by equalsJeffH** [comment]
we need an implementation considerations section and this ought to be addressed in there.

**Comment by bifurcation** [comment]
Seems like it should be addressed by some security considerations / implementation considerations, e.g., recommending that authenticators generate fresh keys for every makeAssertion.

**Comment by nadalin** [comment]
@selfissued can you look at this ?

**Comment by emlun** [comment]
The changes proposed in #672 should go some way to address this.

- If an authenticator supports user verification, it must be used when creating a credential.
- When making an assertion with no `allowCredentials`, user verification must be performed and the authenticator will not give up any information about credentials owned by other users.
- User verification is not required if `allowCredentials` is given, but then the RP must already know the credential ID to use.

**Comment by emlun** [comment]
So it will be possible for Alice to detect whether the authenticator contains Bob's credential `C` only if Alice knows `C`'s credential ID and RP ID.

**Comment by equalsJeffH** [comment]
@emlun: Hm, I'm guessing that the spec has changed out from under your assertions above (https://github.com/w3c/webauthn/issues/96#issuecomment-343227613)... 
> If an authenticator supports user verification, it must be used when **creating a credential**.

Does that hold if the RP has set `options.authenticatorSelection.userVerification` to "discouraged" ?  See step 19.2 in [5.1.3 #createCredential](https://w3c.github.io/webauthn/#createCredential)

> When **making an assertion with no `allowCredentials`**, user verification must be performed and the authenticator will not give up any information about credentials owned by other users.

I am not sure this is presently the case in webauthn's [6.2.3 authenticatorGetAssertion operation](https://w3c.github.io/webauthn/#op-get-assertion), because it appears [IIUC](https://en.wiktionary.org/wiki/IIUC), that in step 19.2 in [5.1.3 #createCredential](https://w3c.github.io/webauthn/#createCredential) an RP can cause |userVerification| to be false, and then in webauthn's the [authenticatorGetAssertion operation](https://w3c.github.io/webauthn/#op-get-assertion), in steps 4 & 5 all creds on the authnr are filtered by only `rpId`.

Similarly, [CTAP says](https://fidoalliance.org/specs/fido-v2.0-ps-20170927/fido-client-to-authenticator-protocol-v2.0-ps-20170927.html#authenticatorGetAssertion):
> If an allowList is not present, locate all credentials that are present on this authenticator and bound to the specified rpId.

[the CTAP editor's draft appears to have the same behavior]

Also, in webauthn's [6.2.3 authenticatorGetAssertion operation](https://w3c.github.io/webauthn/#op-get-assertion), in step 17.4, when `options.allowCredentials` is empty (down at the step bottom), the Note says:
> Note: In this case, the Relying Party did not supply a list of acceptable credential descriptors. Thus, the authenticator is being asked to exercise any credential it may possess that is bound to the Relying Party, as identified by rpId.

WRT your third point:
> User verification is not required if allowCredentials is given, but then the RP must already know the credential ID to use.

Presently, the value of `userVerification` (that is subsequently passed to the authenticator(s)) is set in webauthn's [6.2.3 authenticatorGetAssertion operation](https://w3c.github.io/webauthn/#op-get-assertion), in step 17.2 and appears to be independent of whether `options.allowCredentials` is empty or not. 

Given all this, I'm thinking that @levangongPayPal's [original](https://github.com/w3c/webauthn/issues/96#issue-154577449)..
> Could we ... recommend an implementation approach that would allow to link a credential to an OS account?

..is the way to go, which begs the question of having an "implementation considerations" section(s), along with the other issues we've labeled with [subtype:impl-cons](https://github.com/w3c/webauthn/issues?q=is%3Aopen+is%3Aissue+label%3Asubtype%3Aimpl-cons).

for this we could possibly presently stipulate something like:
> In the case of [=authenticators=] [=bound authenticators|bound=] to platforms having a multiuser operating system, the authenticator and the platform SHOULD work together such that a platform user's webauthn credentials are not visible to other platform users.

WDYT?

**Comment by equalsJeffH** [comment]
@emlun -- are you willing to take on assignment of this issue and composing a PR for it?

**Comment by emlun** [comment]
@equalsJeffH I am.

**Comment by emlun** [comment]
@equalsJeffH @levangongPayPal PR #989 opened. I added this to the Privacy Considerations section as it seemed like an equally good fit as a (new) Implementation Considerations section.

**Comment by equalsJeffH** [comment]
thx @emlun !
